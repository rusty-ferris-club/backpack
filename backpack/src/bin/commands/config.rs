use anyhow::Result as AnyResult;
use backpack::config::Config;
use clap::{Arg, ArgMatches, Command};
use std::path::Path;

pub fn command() -> Command<'static> {
    Command::new("config")
        .about("Create a personal configuration")
        .arg(
            Arg::new("init")
                .short('i')
                .long("init")
                .help("Initialize an empty configuration file")
                .takes_value(false),
        )
        .arg(
            Arg::new("local")
                .long("local")
                .help("Initialize local configuration")
                .takes_value(false),
        )
}

fn print_path(kind: &str, path: &Path) {
    println!(
        "{} {} {}",
        kind,
        if path.exists() { "(found):" } else { "(none):" },
        path.display()
    );
}
pub fn run(_matches: &ArgMatches, subcommand_matches: &ArgMatches) -> AnyResult<bool> {
    if subcommand_matches.is_present("init") {
        let generated = if subcommand_matches.is_present("local") {
            Config::init_local()?
        } else {
            Config::init_global()?
        };
        println!("wrote: {}.", generated.display());
    } else {
        let local = Config::local_config_file();
        let global = Config::global_config_file()?;
        print_path("global", global.as_path());
        print_path("local", local.as_path());

        let t = Config::load_or_default()?.0.to_text()?;
        println!("{}", t);
    }

    Ok(true)
}
