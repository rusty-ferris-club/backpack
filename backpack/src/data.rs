use anyhow::Result as AnyResult;
use core::fmt::Debug;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Clone, Debug)]
pub struct Location {
    pub url: String,
    pub domain: String,
    pub path: String,
    pub project: String,
    pub port: Option<u16>,
    pub query: Option<String>,
    pub subfolder: Option<String>,
    pub gref: Option<String>,
    pub username: String,
    pub password: Option<String>,
    pub is_git: bool,
}
impl Location {
    pub fn git_url(&self) -> String {
        format!(
            "git@{}:{}.git",
            self.domain,
            self.path.trim_start_matches('/')
        ) // needs: auth, port
    }

    pub fn web_url(&self) -> String {
        format!("https://{}{}", self.domain, self.path) // needs: auth, port
    }

    /// Create a logical `Location` from a `Url`.
    ///
    /// # Errors
    ///
    /// This function will return an error if `url` is illegal
    pub fn from(url: &Url, is_git: bool) -> AnyResult<Self> {
        let path = url.path();
        let parts = path.split("/-/").collect::<Vec<_>>();
        let (path, subfolder) = if parts.len() == 2 {
            (parts[0], Some(parts[1].to_string()))
        } else {
            (path, None)
        };

        let project = path
            .split('/')
            .last()
            .ok_or_else(|| anyhow::anyhow!("cannot find project name"))?;

        Ok(Self {
            url: url.to_string(),
            domain: url.domain().unwrap_or_default().to_string(),
            path: path.trim_end_matches(".git").to_string(),
            project: project.to_string(),
            port: url.port(),
            query: url.query().map(str::to_string),
            subfolder,
            gref: url.fragment().map(str::to_string),
            username: url.username().to_string(),
            password: url.password().map(str::to_string),
            is_git,
        })
    }
}

#[derive(Debug)]
pub struct Archive {
    pub url: String,
    pub root: ArchiveRoot,
}

#[derive(Debug)]
pub enum ArchiveRoot {
    Folder(String),
    FirstFolder,
    None,
}
#[derive(Debug)]
pub struct Assets {
    pub archive: Option<Archive>,
    pub git: Option<String>,
}

pub enum Overwrite {
    Ask,
    Always,
    Never,
    Custom(Box<dyn Fn(&str) -> bool>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum CopyMode {
    #[serde(rename = "new")]
    Copy,
    #[serde(rename = "apply")]
    Apply,
    #[serde(rename = "all")]
    All,
}

#[allow(clippy::struct_excessive_bools)]
#[derive(Debug)]
pub struct Opts {
    pub show_progress: bool,
    pub overwrite: bool,
    pub mode: CopyMode,
    pub is_git: bool,
    pub no_cache: bool,
    pub always_yes: bool,
    pub remote: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_location_urls() {
        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            true,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc =
            Location::from(&Url::parse("https://github.com/user/repo").unwrap(), false).unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            false,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");
    }
}
