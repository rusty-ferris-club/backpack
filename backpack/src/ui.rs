use crate::config::Config;
use crate::data::{CopyMode, Opts};
use crate::run::RunnerEvents;
use crate::templates::CopyResult;
use anyhow::{anyhow, Context, Result as AnyResult};
use console::style;
use interactive_actions::data::ActionResult;
use requestty::{Answer, Answers, OnEsc, Question};
use requestty_ui::backend::{Size, TestBackend};
use requestty_ui::events::{KeyEvent, TestEvents};
use std::path::Path;
use std::path::PathBuf;
use std::vec::IntoIter;

pub struct Prompt<'a> {
    config: &'a Config,
    events: Option<TestEvents<IntoIter<KeyEvent>>>,
    show_progress: bool,
}

impl<'a> Prompt<'a> {
    pub fn build(config: &'a Config, show_progress: bool, events: Option<&RunnerEvents>) -> Self {
        events
            .and_then(|evs| evs.prompt_events.as_ref())
            .map_or_else(
                || Prompt::new(config, show_progress),
                |evs| Prompt::with_events(config, evs.clone()),
            )
    }

    pub fn with_events(config: &'a Config, events: Vec<KeyEvent>) -> Self {
        Self {
            config,
            events: Some(TestEvents::new(events)),
            show_progress: false,
        }
    }

    pub fn new(config: &'a Config, show_progress: bool) -> Self {
        Self {
            config,
            events: None::<TestEvents<IntoIter<KeyEvent>>>,
            show_progress,
        }
    }

    /// Fill in missing arguments with a wizard
    ///
    /// # Errors
    ///
    /// This function will return an error if prompting fails
    pub fn fill_missing(
        &mut self,
        shortlink: Option<&str>,
        dest: Option<&str>,
        opts: &Opts,
    ) -> AnyResult<(String, Option<String>, bool)> {
        match (shortlink, dest) {
            (Some(s), Some(d)) => Ok((s.to_string(), Some(d.to_string()), false)),
            (None, d) => {
                let shortlink = if let Some(project) = self.pick_project(&opts.mode)? {
                    project
                } else {
                    self.input_shortlink()?
                };
                if let Some(d) = d {
                    Ok((shortlink, Some(d.to_string()), true))
                } else {
                    Ok((shortlink, self.input_dest()?, true))
                }
            }
            (Some(s), None) => Ok((s.to_string(), self.input_dest()?, true)),
        }
    }

    /// Returns the pick project of this [`Prompt`].
    ///
    /// # Errors
    ///
    /// This function will return an error if interaction is killed
    pub fn pick_project(&self, mode: &CopyMode) -> AnyResult<Option<String>> {
        // accept free input, user wants to input a shortlink directly
        // display where each project comes from
        // impl pick dest, where we do a "my-project" and 1,2,3,4 if exists.
        // add a final confirmation with the data
        // move all UI stuff into prompt
        match self.config.projects_for_selection(Some(mode.clone())) {
            projects if !projects.is_empty() => {
                let options = projects
                    .iter()
                    .map(|(k, p)| {
                        format!(
                            "{} ({})",
                            k,
                            if CopyMode::Apply == p.mode {
                                "apply"
                            } else {
                                "new"
                            }
                        )
                    })
                    .collect::<Vec<_>>();

                let question = Question::select("question")
                    .message("Project (esc for shortlink)")
                    .on_esc(OnEsc::SkipQuestion)
                    .choices(&options)
                    .build();

                let selection = requestty::prompt(vec![question])?
                    .get("question")
                    .and_then(|a| a.as_list_item().cloned());

                match selection {
                    Some(s) if s.index < options.len() => {
                        Ok(projects.get(s.index).map(|(k, _)| (*k).to_string()))
                    }
                    _ => Ok(None),
                }
            }
            _ => Ok(None),
        }
    }

    /// Shows the list of projects
    pub fn show_projects(&self, mode: Option<CopyMode>) {
        println!("Current projects:");
        match self.config.projects_for_selection(mode) {
            projects if !projects.is_empty() => {
                for (name, project) in projects {
                    println!(
                        "- {} ({})",
                        style(name).yellow(),
                        style(&project.shortlink).dim()
                    );
                }
            }
            _ => {
                println!("You have no projects yet.");
            }
        };
    }

    /// Returns the input shortlink of this [`Prompt`].
    ///
    /// # Errors
    ///
    /// This function will return an error if interaction is killed
    pub fn ask_for_project_name(&mut self, repo: &str) -> AnyResult<String> {
        let question = Question::input("question")
            .validate(|v, _| {
                if v.is_empty() {
                    Err("cannot be empty".into())
                } else {
                    Ok(())
                }
            })
            .message(format!("A name for '{repo}'?"))
            .build();
        let name = self
            .prompt_one(question)?
            .as_string()
            .ok_or_else(|| anyhow::anyhow!("cannot parse input"))?
            .to_string();
        Ok(name)
    }

    /// Returns the input shortlink of this [`Prompt`].
    ///
    /// # Errors
    ///
    /// This function will return an error if interaction is killed
    pub fn input_shortlink(&mut self) -> AnyResult<String> {
        let question = Question::input("question")
            .validate(|v, _| {
                if v.is_empty() {
                    Err("cannot be empty".into())
                } else {
                    Ok(())
                }
            })
            .message("Shortlink")
            .build();
        let shortlink = self
            .prompt_one(question)?
            .as_string()
            .ok_or_else(|| anyhow::anyhow!("cannot parse input"))?
            .to_string();
        Ok(shortlink)
    }

    /// Returns the input shortlink of this [`Prompt`].
    ///
    /// # Errors
    ///
    /// This function will return an error if interaction is killed
    pub fn input_dest(&mut self) -> AnyResult<Option<String>> {
        let b = Question::input("question").message("Destination");
        let question = b.build();

        let dest = self
            .prompt_one(question)?
            .as_string()
            .ok_or_else(|| anyhow::anyhow!("cannot parse input"))?
            .to_string();

        match dest {
            s if s.is_empty() => Ok(None),
            s => Ok(Some(s)),
        }
    }

    /// Returns the input shortlink of this [`Prompt`].
    ///
    /// # Errors
    ///
    /// This function will return an error if interaction is killed
    pub fn are_you_sure(&mut self, text: &str) -> AnyResult<bool> {
        let question = Question::confirm("question")
            .message(text)
            .default(true)
            .build();

        Ok(self.prompt_one(question)?.as_bool().unwrap_or(false))
    }

    /// Confirm file overwrite
    ///
    /// # Errors
    ///
    /// This function will return an error if prompting does not work
    pub fn confirm_overwrite(&mut self, file: &Path) -> AnyResult<bool> {
        let question = Question::confirm("question")
            .message(format!("'{}' already exists. overwrite?", file.display()))
            .build();

        Ok(self.prompt_one(question)?.as_bool().unwrap_or(false))
    }

    pub fn say_resolving(&self) {
        if self.show_progress {
            println!("üîÆ Resolving...");
        }
    }

    pub fn say_fetching(&self) {
        if self.show_progress {
            println!("üöö Fetching content...");
        }
    }

    pub fn say_unpacking(&self) {
        if self.show_progress {
            println!("üéí Unpacking files...");
        }
    }

    pub fn say_action(&self, name: &str) {
        println!("{name}");
    }

    pub fn say_actions(&self, number: usize) {
        if self.show_progress {
            println!("üçø Running {number} action(s):");
        }
    }

    pub fn say_done(&self, res: &[CopyResult], maybe_actions: Option<&Vec<ActionResult>>) {
        if self.show_progress {
            let total = res.len();
            let total_actions = maybe_actions.map_or(0, Vec::len);
            let cutoff = 10;

            if let Some(ars) = maybe_actions {
                println!();
                for ar in ars.iter() {
                    println!(" {} {}", style("‚úÖ").green(), style(&ar.name).dim());
                }
                println!();
            };

            res.iter().enumerate().take(cutoff).for_each(|(i, s)| {
                if i == 0 {
                    println!();
                }
                println!(
                    " {} {} {}",
                    style("+").green(),
                    style(s.dest.display()).dim(),
                    style(format!("{:?}", s.op)).yellow().dim()
                );
            });
            if total > cutoff {
                println!(
                    "   {} {} {}",
                    style("... and").dim(),
                    style(total - cutoff).yellow(),
                    style("more file(s).").dim()
                );
            }
            println!(
                "\nüéâ Done: {} file(s) copied with {} action(s).",
                style(total).yellow(),
                style(total_actions).yellow()
            );
        }
    }

    pub fn say(&self, text: &str) {
        println!("{text}");
    }

    /// Ask if user wants to edit a file and open editor
    ///
    /// # Errors
    ///
    /// This function will return an error if IO failed
    pub fn suggest_edit(&mut self, text: &str, path: &Path) -> AnyResult<()> {
        if self.are_you_sure(text)? {
            edit::edit_file(path)?;
        }
        Ok(())
    }

    fn prompt_one<I: Into<Question<'a>>>(&mut self, question: I) -> AnyResult<Answer> {
        match self.events {
            Some(ref mut events) => {
                let mut backend = TestBackend::new(Size::from((50, 20)));
                requestty::prompt_one_with(question, &mut backend, events)
                    .context("cannot take input")
            }
            None => requestty::prompt_one(question).context("cannot take input"),
        }
    }

    /// Prompt questions
    ///
    /// # Errors
    ///
    /// This function will return an error if prompting fails
    pub fn prompt<Q>(&mut self, questions: Q) -> AnyResult<Answers>
    where
        Q: IntoIterator<Item = Question<'a>>,
    {
        match self.events {
            Some(ref mut events) => {
                let mut backend = TestBackend::new(Size::from((50, 20)));
                requestty::prompt_with(questions, &mut backend, events).context("cannot take input")
            }
            None => requestty::prompt(questions).context("cannot take input"),
        }
    }
}

/// Guess a destination name by generating a "my-projectN..99" for the user.
///
/// # Errors
///
/// This function will return an error if too many projects have been tried
pub fn guess_dest() -> AnyResult<String> {
    Ok((1..99)
        .map(|idx| PathBuf::from(format!("my-project{idx}")))
        .find(|p| !p.exists())
        .ok_or_else(|| anyhow!("too many apps generated, pick a custom app name?"))?
        .display()
        .to_string())
}
