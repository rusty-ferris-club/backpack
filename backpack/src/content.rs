use crate::config::{Swap, SwapKind};
use crate::data::{CopyMode, Location, Opts, Overwrite};
use crate::ui::Prompt;
use anyhow::{bail, Result};
use interactive_actions::{
    data::ActionResult,
    data::{Action, ActionHook},
    ActionRunner,
};
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use tera::Tera;
use tracing::warn;
use walkdir;

pub struct Coordinate<'a> {
    pub source: &'a Path,
    pub dest: Option<&'a Path>,
    pub location: &'a Location,
    pub remove_source: bool,
}

#[derive(Default)]
pub struct Deployer {}

impl Deployer {
    #[tracing::instrument(skip_all, err)]
    pub fn deploy(
        &self,
        coord: Coordinate<'_>,
        mut action_runner: Option<ActionRunner<'_>>,
        swaps: Option<&Vec<Swap>>,
        vars: &mut BTreeMap<String, String>,
        opts: &Opts,
        prompt: &mut Prompt<'_>,
    ) -> Result<(Vec<String>, Option<Vec<ActionResult>>)> {
        // xxx: either way canonicalize paths.
        let final_source = coord
            .source
            .join(coord.location.subfolder.clone().unwrap_or_default());
        let dest = coord.dest.map(std::path::Path::to_path_buf);
        let location_path = coord.location.subfolder.clone().map(PathBuf::from);

        // is this "deploying" a single file or a folder?
        let is_file = final_source.is_file();

        let final_dest = if is_file {
            // final dest = dest | location+fname | fname
            let fname = final_source
                .file_name()
                .ok_or_else(|| anyhow::anyhow!("cannot get file name for {:?}", final_source))?;
            dest.or_else(|| {
                location_path.and_then(|loc| loc.parent().map(|p| p.to_path_buf().join(fname)))
            })
            .unwrap_or_else(|| PathBuf::from(fname))
        } else {
            dest.or(location_path)
                .unwrap_or_else(|| PathBuf::from(".".to_string()))
        };

        let actions_dest = if is_file {
            final_dest.parent().unwrap_or_else(|| Path::new("."))
        } else {
            final_dest.as_path()
        };

        let before_actions = if let Some(action_runner) = action_runner.as_mut() {
            Some(action_runner.run(
                Some(actions_dest),
                vars,
                ActionHook::Before,
                Some(|action: &Action| prompt.say_action(action.name.as_str())),
            )?)
        } else {
            None
        };

        let swapper = Swapper::with_vars(swaps, vars)?;
        let files = match opts.mode {
            CopyMode::Copy => {
                if final_dest.exists() {
                    anyhow::bail!("path already exists: {}", final_dest.display());
                }
                self.copy(
                    &swapper,
                    &final_source,
                    &final_dest,
                    is_file,
                    Overwrite::Always,
                    prompt,
                )?
            }
            CopyMode::Apply => self.copy(
                &swapper,
                &final_source,
                &final_dest,
                is_file,
                if opts.overwrite {
                    Overwrite::Always
                } else {
                    Overwrite::Ask
                },
                prompt,
            )?,
            CopyMode::All => {
                vec![]
            }
        };
        if coord.remove_source {
            // xxx don't remove for now
            warn!(
                "remove requested, but not removing '{}'",
                coord.source.display()
            );
        }

        let after_actions = if let Some(action_runner) = action_runner.as_mut() {
            Some(action_runner.run(
                Some(actions_dest),
                vars,
                ActionHook::After,
                Some(|action: &Action| prompt.say_action(action.name.as_str())),
            )?)
        } else {
            None
        };

        // just a fancy zip + fold of the two collections
        let actions = [before_actions, after_actions]
            .into_iter()
            .flatten()
            .reduce(|mut a, mut b| {
                a.append(&mut b);
                a
            });

        Ok((files, actions))
    }

    #[tracing::instrument(skip_all, err)]
    fn copy(
        &self,
        swapper: &Swapper,
        source: &Path,
        dest: &Path,
        is_file: bool,
        overwrite: Overwrite,
        prompt: &mut Prompt<'_>,
    ) -> Result<Vec<String>> {
        // swapfs = Swaps::new
        if is_file {
            let swapped = swapper.copy_to(source, dest)?;
            return Ok(vec![swapped.display().to_string()]);
        }

        let mut copied = vec![];
        walkdir::WalkDir::new(source)
            .into_iter()
            .try_for_each(|entry| {
                let entry = entry?;
                let path = entry.path();
                if path.is_file() {
                    let to = dest.join(path.strip_prefix(source)?);
                    let to_path = to.as_path();

                    //
                    // trading off some inefficiency for readability here.
                    // the swapped path can be created only once, but instead it's created
                    // many times below in `exists`, in prompt, and by each of the `copy`s
                    // instead of creating a swapped path and *remembering* to pass throughout the workflow,
                    // we hide its creation.
                    //
                    // in addition, each `copy` will check to see if the parent exists, and if not, create it.
                    // we could create the parent just once here, but again - hiding it inside swapper is better
                    // than remembering nuances in the prices of checking parent folder more times than needed.
                    //
                    if swapper.exists(to_path) {
                        let should_copy = match overwrite {
                            Overwrite::Always => true,
                            Overwrite::Ask => prompt
                                .confirm_overwrite(swapper.path(to_path).as_path())
                                .unwrap_or(false),
                            _ => false,
                        };
                        if should_copy {
                            let to_swapped = swapper.copy_to(path, to_path)?;
                            copied.push(to_swapped.display().to_string());
                        }
                    } else {
                        let to_swapped = swapper.copy_to(path, to_path)?; //swapfs.copy(..)
                        copied.push(to_swapped.display().to_string());
                    }
                }

                anyhow::Ok(())
            })?;
        Ok(copied)
    }
}

pub struct Swapper {
    swaps: Vec<Swap>,
}
impl Swapper {
    ///
    /// Create a swapper with fully populated swaps
    ///
    /// # Errors
    /// Return errors when swaps cannot be populated, e.g. when a `val_template` is illegal
    #[must_use]
    pub fn with_vars(swaps: Option<&Vec<Swap>>, vars: &BTreeMap<String, String>) -> Result<Self> {
        let empty = vec![];
        let s = swaps.unwrap_or(&empty);
        Ok(Self {
            swaps: render_swaps(s, vars)?,
        })
    }

    pub fn path(&self, p: &Path) -> PathBuf {
        let pstr = p.display().to_string();
        let mut s = pstr.clone();
        for swap in &self.swaps {
            if match_path(pstr.as_str(), swap) {
                if let Some(val) = swap.val.as_ref() {
                    s = s.replace(swap.key.as_str(), val);
                }
            }
        }
        PathBuf::from(s)
    }
    pub fn copy_to(&self, source: &Path, dest: &Path) -> Result<PathBuf> {
        let swapped = self.path(dest);
        let parent = dest
            .parent()
            .ok_or_else(|| anyhow::anyhow!("cannot get parent for {:?}", dest))?;
        if !parent.exists() {
            fs::create_dir_all(&parent)?;
        };
        fs::copy(source, &swapped)?; // swapfs.copy(s, d)
        Ok(swapped)
    }
    pub fn exists(&self, dest: &Path) -> bool {
        let p = self.path(dest);
        p.exists()
    }
}

fn render_swaps(swaps: &[Swap], varbag: &BTreeMap<String, String>) -> Result<Vec<Swap>> {
    let mut tera = Tera::default();
    tera_text_filters::register_all(&mut tera);
    let context = tera::Context::from_serialize(varbag)?;
    swaps
        .iter()
        .map(|swap| {
            let val = match (swap.val.as_ref(), swap.val_template.as_ref()) {
                (Some(v), _) => v.clone(),
                (None, Some(v)) => tera.render_str(v, &context)?,
                (None, None) => bail!("each swap should have either `val` or `val_template`"),
            };
            let mut s = swap.clone();
            s.val = Some(val);
            Ok(s)
        })
        .collect::<Result<Vec<_>>>()
}

fn match_path(p: &str, swap: &Swap) -> bool {
    match swap.kind {
        SwapKind::Path | SwapKind::All => swap.path.as_ref().map_or(true, |exp| exp.is_match(p)),
        SwapKind::Content => false,
    }
}

fn match_copy(p: &str, swap: &Swap) -> bool {
    match swap.kind {
        SwapKind::Content | SwapKind::All => swap.path.as_ref().map_or(true, |exp| exp.is_match(p)),
        SwapKind::Path => false,
    }
}

#[cfg(test)]
mod tests {
    use regex::Regex;
    use std::vec;

    use insta::assert_yaml_snapshot;
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn test_render_swaps() {
        let mut h = BTreeMap::new();
        h.insert("world".to_string(), "crewl world".to_string());

        let swaps = render_swaps(
            vec![Swap {
                key: "key".to_string(),
                val_template: Some("Hello {{world}}".to_string()),
                val: None,
                ..Default::default()
            }]
            .as_slice(),
            &h,
        )
        .unwrap();
        assert_yaml_snapshot!(swaps);
    }

    #[test]
    fn test_render_swaps_inflections() {
        let mut h = BTreeMap::new();
        h.insert("world".to_string(), "crewl world".to_string());

        let swaps = render_swaps(
            vec![Swap {
                key: "key".to_string(),
                val_template: Some("Hello {{world | kebab_case}}".to_string()),
                val: None,
                ..Default::default()
            }]
            .as_slice(),
            &h,
        )
        .unwrap();
        assert_yaml_snapshot!(swaps);
    }

    #[test]
    fn test_render_swaps_empty_context() {
        let h = BTreeMap::new();

        let swaps = render_swaps(
            vec![Swap {
                key: "key".to_string(),
                val_template: Some("Hello {{world}}".to_string()),
                val: None,
                ..Default::default()
            }]
            .as_slice(),
            &h,
        );
        assert_eq!(
            swaps.unwrap_err().to_string(),
            "Failed to render '__tera_one_off'"
        );
    }

    #[test]
    fn test_render_wrong_context() {
        let mut h = BTreeMap::new();
        h.insert("foobar".to_string(), "crewl world".to_string());

        let swaps = render_swaps(
            vec![Swap {
                key: "key".to_string(),
                val_template: Some("Hello {{world}}".to_string()),
                val: None,
                ..Default::default()
            }]
            .as_slice(),
            &h,
        );

        assert_eq!(
            swaps.unwrap_err().to_string(),
            "Failed to render '__tera_one_off'"
        );
    }

    #[test]
    fn test_match_path() {
        assert!(match_path("some/path", &Swap::default()));
        assert!(match_path(
            "some/path",
            &Swap {
                kind: SwapKind::All,
                ..Default::default()
            }
        ));
        assert!(match_path(
            "some/path",
            &Swap {
                kind: SwapKind::Path,
                ..Default::default()
            }
        ));
        assert!(!match_path(
            "some/path",
            &Swap {
                kind: SwapKind::Content,
                ..Default::default()
            }
        ));
        assert!(!match_path(
            "some/path",
            &Swap {
                kind: SwapKind::Path,
                path: Some(Regex::new(".*foo").unwrap()),
                ..Default::default()
            }
        ));
        assert!(match_path(
            "some/path",
            &Swap {
                kind: SwapKind::Path,
                path: Some(Regex::new("some/.*").unwrap()),
                ..Default::default()
            }
        ));
    }

    #[test]
    fn test_match_copy() {
        assert!(match_copy("some/path", &Swap::default()));
        assert!(match_copy(
            "some/path",
            &Swap {
                kind: SwapKind::All,
                ..Default::default()
            }
        ));
        assert!(!match_copy(
            "some/path",
            &Swap {
                kind: SwapKind::Path,
                ..Default::default()
            }
        ));
        assert!(match_copy(
            "some/path",
            &Swap {
                kind: SwapKind::Content,
                ..Default::default()
            }
        ));
        assert!(!match_copy(
            "some/path",
            &Swap {
                kind: SwapKind::Content,
                path: Some(Regex::new(".*foo").unwrap()),
                ..Default::default()
            }
        ));
        assert!(match_copy(
            "some/path",
            &Swap {
                kind: SwapKind::Content,
                path: Some(Regex::new("some/.*").unwrap()),
                ..Default::default()
            }
        ));
    }

    #[test]
    fn test_swap_path() {
        let swaps = vec![Swap {
            key: "$SWAPME$".to_string(),
            kind: SwapKind::All,
            val_template: Some("{{greeting | kebab_case}}".to_string()),
            ..Default::default()
        }];
        let mut vars = BTreeMap::new();
        vars.insert("greeting".into(), "hello world".into());

        let swapper = Swapper::with_vars(Some(&swaps), &vars).unwrap();
        assert_eq!(
            "some/hello-world/path",
            swapper
                .path(Path::new("some/$SWAPME$/path"))
                .display()
                .to_string()
        );
        assert_eq!(
            "some/naive/path",
            swapper
                .path(Path::new("some/naive/path"))
                .display()
                .to_string()
        );
    }
}
