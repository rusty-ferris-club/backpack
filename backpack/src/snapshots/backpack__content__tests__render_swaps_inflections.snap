---
source: backpack/src/content.rs
expression: swaps
---
- key: key
  val_template: "Hello {{world | kebab_case}}"
  val: Hello crewl-world
  path: ~
  kind: all

